### PR TITLE
UCT: Make uct_iface_release_desc more flexible

### PR DESCRIPTION
**Pull request by brminich**
The main purpose of this PR is to support variable header lengths for the transport

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1411/ for details.

**Comment by brminich** [comment]
@yosefe , @MattBBaker please take a look

**Comment by MattBBaker** [comment]
Had a look, compiles and runs on the UGNI machines. 👍

File: src/uct/api/tl.h [hunk 1]
```diff
@@ -211,17 +209,20 @@ typedef struct uct_ep {
...
-
+struct uct_recv_desc {
+    uct_desc_release_callback_t cb;
+};
 
-#define uct_recv_desc_iface(_desc) \
-    ((((uct_recv_desc_t*)(_desc)) - 1)->iface)
 
+#define uct_recv_desc(_desc) \
+   *((uct_recv_desc_t**)(((uct_recv_desc_t*)(_desc)) - 1))
```
Comment (yosefe) [inline src/uct/api/tl.h]: this casting probably works, but not entirely right. should be:
```
(   *( ( (uct_recv_desc_t**)(_desc) ) - 1)   )
```

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3337/ for details (Mellanox internal link).

File: src/uct/api/tl.h (hunk 1 above)
Comment (brminich) [inline src/uct/api/tl.h]: Right, we have to shift by address, not struct size. Thanks

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1412/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3339/ for details (Mellanox internal link).

**Pull request merged**